GAE/src/org/akvo/flow/rest/DataPointAssignmentRestService.java

```diff
@@ -0,0 +1,181 @@
+        statusDto.setStatus("failed");
+        if (dpa != null) {
+            //Nothing depends on this; just delete
+            dataPointAssignmentDao.delete(dpa);
+            statusDto.setStatus("ok");
+        }
+
+        response.put("meta", statusDto);
+        return response;
+    }
+
+    @RequestMapping(method = RequestMethod.PUT, value = "/{id}")
+    @ResponseBody
+    public Map<String, DataPointAssignmentDto> updateDataPointAssignment(
+            @PathVariable("id")
+            Long id,
+            @RequestBody
+            DataPointAssignmentPayload payload) {
+
+        final DataPointAssignmentDto dto = payload.getData_point_assignment();
+        if (!id.equals(dto.getKeyId())) {
+            throw new HttpMessageNotReadableException("Ids don't match: " + id
+                    + " <> " + dto.getKeyId());
+        }
+        final DataPointAssignment assignment = dataPointAssignmentDao.getByKey(dto.getKeyId());
+        if (assignment == null) {
+            throw new ResourceNotFoundException(
+                    "DataPointAssignment with id: " + dto.getKeyId() + " not found");
+        }
+
+        final DataPointAssignment dpa = marshallToDomain(dto);
+        dataPointAssignmentDao.save(dpa);
+
+        final HashMap<String, DataPointAssignmentDto> response = new HashMap<String, DataPointAssignmentDto>();
+        response.put("data_point_assignment", marshallToDto(dpa));
+
+        return response;
+    }
+
+    @RequestMapping(method = RequestMethod.POST, value = "")
+    @ResponseBody
+    public Map<String, DataPointAssignmentDto> newDataPointAssignment(
+            @RequestBody
+            DataPointAssignmentPayload payload) {
+
+        final DataPointAssignmentDto dto = payload.getData_point_assignment();
+        final DataPointAssignment dpa  = marshallToDomain(dto);
+
+        dataPointAssignmentDao.save(dpa); //fills in new key
+
+        final HashMap<String, DataPointAssignmentDto> response = new HashMap<String, DataPointAssignmentDto>();
+        response.put("data_point_assignment", marshallToDto(dpa));
+        return response;
+    }
+
+    private DataPointAssignmentDto marshallToDto(DataPointAssignment dpa) {
+        final DataPointAssignmentDto dto = new DataPointAssignmentDto();
+
+        BeanUtils.copyProperties(dpa, dto);
+        if (dpa.getKey() != null) {
+            dto.setKeyId(dpa.getKey().getId());
+        }
+        //Lists must be copied separately
+        dto.setDataPointIds(dpa.getDataPointIds());
+
+        return dto;
+    }
+
+    private DataPointAssignment marshallToDomain(DataPointAssignmentDto dto) {
+        final DataPointAssignment dpa = new DataPointAssignment();
+
+        BeanUtils.copyProperties(dto, dpa);
+        if (dto.getKeyId() != null) {
+            dpa.setKey(KeyFactory.createKey("DataPointAssignment", dto.getKeyId()));
+        }
+        //Lists must be copied separately
+        dpa.setDataPointIds(dto.getDataPointIds());
```

Comment:
The `BeanUtils.copyProperties()` cant handle this?

Reply:
This is the way it is done elsewhere., but I will experiment.

Works automagically, too. Cool.

Reply:
May have been done that way just to change field names.